### PR TITLE
Feature/joi

### PR DESCRIPTION
**Pull request by Itay9263**

### CODE REVIEW

File: src/exc.ts [hunk 1]
```diff
@@ -50,6 +50,12 @@ export class CannotPerformOperationError extends AppError {
   }
 }
 
+export class ObjectValidationError extends AppError {
+  constructor(m: string, status = 400) {
+    super(m, status);
+    Object.setPrototypeOf(this, CannotPerformOperationError.prototype);
```
Comment (Hyaxia) [inline src/exc.ts:56]: small copy paste fix - change `CannotPerformOperationError` to `ObjectValidationError`

File: src/api/volunteerRequest.ts [hunk 2]
```diff
@@ -27,9 +29,30 @@ const router = Router();
...
+  .post(async (req: Request, res: Response, next: NextFunction) => {
+    try {
+      const { volunteerRequest } = req.body;
+      res.json(await createVolunteerRequest(volunteerRequest as VolunteerRequest));
+    } catch (e) {
+      next(e);
+    }
+  });
+
+router.put('/:id', async (req: Request, res: Response, next: NextFunction) => {
```
Comment (Hyaxia) [inline src/api/volunteerRequest.ts:50]: add swagger docs to the route

File: src/api/volunteerRequest.ts [hunk 3]
```diff
@@ -27,9 +29,30 @@ const router = Router();
...
+router
+  .route('/')
+  .get(async (req: Request, res: Response, next: NextFunction) => {
+    try {
+      res.json(await getRelevantAndOpenVolunteerRequests());
+    } catch (e) {
+      next(e);
+    }
+  })
+  .post(async (req: Request, res: Response, next: NextFunction) => {
```
Comment (Hyaxia) [inline src/api/volunteerRequest.ts:41]: missing `authMiddleware`

File: src/app/volunteerRequest.ts [hunk 4]
```diff
@@ -15,6 +17,18 @@ export const getVolunteerRequestsOfProgram = async (
...
 
+export const createVolunteerRequest = async (volunteerRequest: VolunteerRequest): Promise<VolunteerRequest> => {
+  return volunteerRequestRepository.save(validateSchema(volunteerRequestSchema, volunteerRequest));
+};
+
+export const updateVolunteerRequest = async (
+  id: number,
+  volunteerRequestInfo: Partial<VolunteerRequest>
+): Promise<UpdateResult> => {
+  if (!id) throw new BadRequestError('Missing Id to update volunteer request by');
```
Comment (Hyaxia) [inline src/app/volunteerRequest.ts:28]: missing checks who can update the volunteer request

File: src/api/volunteerRequest.ts [hunk 5]
```diff
@@ -27,9 +29,30 @@ const router = Router();
...
+    try {
+      const { volunteerRequest } = req.body;
+      res.json(await createVolunteerRequest(volunteerRequest as VolunteerRequest));
+    } catch (e) {
+      next(e);
+    }
+  });
+
+router.put('/:id', async (req: Request, res: Response, next: NextFunction) => {
   try {
```
Comment (Hyaxia) [inline src/api/volunteerRequest.ts:51]: missing `authMiddleware`

File: src/app/volunteerRequest.ts [hunk 6]
```diff
@@ -15,6 +17,18 @@ export const getVolunteerRequestsOfProgram = async (
   return volunteerRequestRepository.requestsOfProgram(programId, institutionId, startDate);
 };
 
+export const createVolunteerRequest = async (volunteerRequest: VolunteerRequest): Promise<VolunteerRequest> => {
+  return volunteerRequestRepository.save(validateSchema(volunteerRequestSchema, volunteerRequest));
```
Comment (Hyaxia) [inline src/app/volunteerRequest.ts:21]: missing checks for creation of the volunteer request, for example, if the user is from institution id 1, then the volunteer request must have also institution id 1. I'm sure there are more validations to make besides that.

File: src/app/schema.validators.ts [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+  email: Joi.string().email(),
+  userType: Joi.string().valid(...Object.values(UserType)),
+  companyId: Joi.number().when('userType', {
+    is: UserType.VOLUNTEER,
+    then: Joi.number(),
+    otherwise: onlyNull
+  }),
+  programId: Joi.number().when('userType', {
+    is: UserType.VOLUNTEER,
+    then: onlyNull,
```
Comment (Hyaxia) [inline src/app/schema.validators.ts:25]: why is it `onlyNull`?

File: src/app/user.ts [hunk 8]
```diff
@@ -49,16 +42,10 @@ export const login = async (userId: string): Promise<loginResponse> => {
 
 export const register = async (user: Partial<User>): Promise<loginResponse> => {
```
Comment (Hyaxia) [inline src/app/user.ts:43]: can we ask the UI to send us a token that will signal that the user has been authenticated by google or something similar? so not every request sent to this route will be processed but only those that have been authentiacted

**Review by Hyaxia — CHANGES_REQUESTED** [review]
besides the other comments, tests for each new use case should be added.
if you want me to help you swagger docs or something else tell me

File: src/api/volunteerRequest.ts [hunk 9]
```diff
@@ -27,9 +29,30 @@ const router = Router();
...
-    res.json(await getRelevantAndOpenVolunteerRequests());
+    const { volunteerRequestInfo } = req.body;
+    const id = +req.params.id;
```
[the remark this answers is not in the document]
Comment (Itay9263) [reply src/api/volunteerRequest.ts:53]: yep

**Review by Hyaxia — APPROVED** [review]

**Pull request merged**